src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -147,30 +147,38 @@ public JsonClassInfo(Type type, JsonSerializerOptions options)
                         }
 
                         JsonPropertyInfo[] cacheArray;
-                        if (DetermineExtensionDataProperty(cache))
+                        if (DetermineExtensionDataProperty(cacheDictionary))
                         {
                             // Remove from cache since it is handled independently.
-                            cache.Remove(DataExtensionProperty!.NameAsString!);
+                            cacheDictionary.Remove(DataExtensionProperty!.NameAsString!);
 
-                            cacheArray = new JsonPropertyInfo[cache.Count + 1];
+                            cacheArray = new JsonPropertyInfo[cacheDictionary.Count + 1];
 
                             // Set the last element to the extension property.
-                            cacheArray[cache.Count] = DataExtensionProperty;
+                            cacheArray[cacheDictionary.Count] = DataExtensionProperty;
                         }
                         else
                         {
-                            cacheArray = new JsonPropertyInfo[cache.Count];
+                            cacheArray = new JsonPropertyInfo[cacheDictionary.Count];
                         }
 
-                        // Set fields when finished to avoid concurrency issues.
-                        PropertyCache = cache;
-                        cache.Values.CopyTo(cacheArray, 0);
+                        // Copy the dictionary cache to the array cache.
+                        cacheDictionary.Values.CopyTo(cacheArray, 0);
+
+                        // Set the array cache field at this point since it is completely initialized.
+                        // It can now be safely accessed by other threads.
                         PropertyCacheArray = cacheArray;
 
+                        // Allow constructor parameter logic to remove items from the dictionary since the JSON
+                        // property values will be passed to the constructor and do not call a property setter.
                         if (converter.ConstructorIsParameterized)
                         {
-                            InitializeConstructorParameters(converter.ConstructorInfo!);
+                            InitializeConstructorParameters(cacheDictionary, converter.ConstructorInfo!);
```

Comment:
Changed this since we should pass in the dictionary cache variable instead of the field so no other thread will see the field until it is ready.